src/gazebo_vision_plugin.cpp

```diff
@@ -113,21 +143,178 @@ void VisionPlugin::Load(physics::ModelPtr model, sdf::ElementPtr sdf)
 }
 
-void VisionPlugin::OnUpdate(const common::UpdateInfo&)
-{
+double VisionPlugin::calcTimeStep() {
 #if GAZEBO_MAJOR_VERSION >= 9
-  common::Time current_time = _world->SimTime();
+  _current_time = _world->SimTime();
 #else
-  common::Time current_time = _world->GetSimTime();
+  _current_time = _world->GetSimTime();
 #endif
-  double dt = (current_time - _last_pub_time).Double();
+  return (_current_time - _last_pub_time).Double();
+}
+
+#if BUILD_ROS_INTERFACE == 1
+void VisionPlugin::handle_vision_data(double sec,
+                                      geometry_msgs::Point pos,
+                                      geometry_msgs::Quaternion att,
+                                      geometry_msgs::Vector3 lin_vel,
+                                      geometry_msgs::Vector3 ang_vel,
+                                      double (&pose_cov)[36],
+                                      double (&twist_cov)[36])
+{
+   // Fill odom msg
+    odom_msg.set_usec(sec * 1e6); // convert sec to usec
+
+    gazebo::msgs::Vector3d* position = new gazebo::msgs::Vector3d();
+    position->set_x(pos.x);
+    position->set_y(pos.y);
+    position->set_z(pos.z);
+    odom_msg.set_allocated_position(position);
+
+    gazebo::msgs::Quaternion* orientation = new gazebo::msgs::Quaternion();
+    orientation->set_x(att.x);
+    orientation->set_y(att.y);
+    orientation->set_z(att.z);
+    orientation->set_w(att.w);
+    odom_msg.set_allocated_orientation(orientation);
+
+    gazebo::msgs::Vector3d* linear_velocity = new gazebo::msgs::Vector3d();
+    linear_velocity->set_x(lin_vel.x);
+    linear_velocity->set_y(lin_vel.y);
+    linear_velocity->set_z(lin_vel.z);
+    odom_msg.set_allocated_linear_velocity(linear_velocity);
+
+    gazebo::msgs::Vector3d* angular_velocity = new gazebo::msgs::Vector3d();
+    angular_velocity->set_x(ang_vel.x);
+    angular_velocity->set_y(ang_vel.y);
+    angular_velocity->set_z(ang_vel.z);
+    odom_msg.set_allocated_angular_velocity(angular_velocity);
+
+    for (int i = 0; i < 36; i++) {
+      odom_msg.add_pose_covariance(pose_cov[i]);
+      odom_msg.add_twist_covariance(twist_cov[i]);
+    }
+
+    _last_pub_time = _current_time;
+
+    // publish odom msg
+    _pub_odom->Publish(odom_msg);
+}
+
+void VisionPlugin::rosPoseCovCallBack(const geometry_msgs::PoseWithCovarianceStampedConstPtr& odomMsg)
+{
+  double dt = calcTimeStep();
+
+  if (dt > 1.0 / _pub_rate) {
+
+    geometry_msgs::Vector3 lin_vel;
+    lin_vel.x = lin_vel.y = lin_vel.z = 0; // linear velocity is not provided by sensor
+    geometry_msgs::Vector3 ang_vel;
+    ang_vel.x = ang_vel.y = ang_vel.z = 0; // angular velocity is not provided by sensor
+
+    double pose_covariance[36] = {};
+    double twist_covariance[36] = {}; // all values are zero
+    memcpy(pose_covariance, odomMsg->pose.covariance.data(), sizeof(double)*36);
+
+    handle_vision_data(odomMsg->header.stamp.toSec(),
```

Comment:
`odomMsg->header.stamp.toNSec(),`

Reply:
Not sure. https://github.com/PX4/sitl_gazebo/blob/master/msgs/Odometry.proto#L8 shows that timestamp is in micosecs, while `odomMsg->header.stamp.toNSec()` gives nanosecs.

Reply:
@EliaTarasov you didn't change `toSec()` to `toNSec()`. If use it, you don't have to do the conversion on the method above.

Reply:
@TSC21 can you please clarify my comment above?

Reply:
Yes. The idea is to change this and propate the timestamp in nanosecs. Check https://github.com/PX4/sitl_gazebo/pull/223#discussion_r198786207.